logexp/cmd/main.go

```diff
@@ -0,0 +1,257 @@
+	nodeLogs = []string{"kube-proxy", "node-problem-detector", "fluentd"}
+
+	// Cloud provider specific logfiles.
+	awsLogs = []string{"cloud-init-output"}
+	gceLogs = []string{"startupscript"}
+	kubemarkLogs = []string{}
+
+	// System services/kernel related logfiles.
+	kernelLog = "kern"
+	initdLogs = []string{"docker"}
+	supervisordLogs = []string{"kubelet", "supervisor/supervisord", "supervisor/kubelet-stdout", "supervisor/kubelet-stderr", "supervisor/docker-stdout", "supervisor/docker-stderr"}
+	systemdServices = []string{"kern", "kubelet", "docker"}
+	systemdSetupServices = []string{"kube-node-installation", "kube-node-configuration"}
+	nodeSystemdServices = []string{"node-problem-detector"}
+)
+
+func registerFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&nodeName, "node-name", "", "Name of the node this log exporter is running on")
+	fs.StringVar(&gcsArtifactsPath, "gcs-artifacts-path", "", "Path to the GCS artifacts dir under which to upload logs, for eg: gs://my-logs-bucket/logs")
+	fs.StringVar(&kubernetesProvider, "kubernetes-provider", "", "Cloud provider for this node (gce/gke/aws/kubemark/..)")
+	fs.StringVar(&gcloudAuthFilePath, "gcloud-auth-file-path", "/etc/service-account/service-account.json", "Path to gcloud service account file, for authenticating gsutil to write to GCS bucket")
+	fs.BoolVar(&enableHollowNodeLogs, "enable-hollow-node-logs", false, "Enable uploading hollow node logs too. Relevant only for kubemark nodes")
+	fs.DurationVar(&sleepDuration, "sleep-duration", 60*time.Second, "Duration to sleep before exiting with success. Useful for making pods schedule with hard anti-affinity when run as a job on a k8s cluster")
+}
+
+// Check if the config provided through the flags take valid values.
+func checkConfigValidity() error {
+	glog.Infof("Verifying if a valid config has been provided through the flags")
+	if nodeName == "" {
+		return fmt.Errorf("Flag --node-name has its value unspecified")
+	}
+	if gcsArtifactsPath == "" {
+		return fmt.Errorf("Flag --gcs-artifacts-path has its value unspecified")
+	}
+	if _, err := os.Stat(gcloudAuthFilePath); os.IsNotExist(err) {
+		return fmt.Errorf("Could not find the gcloud service account file: %v", err)
+	} else {
+		cmd := exec.Command("gcloud", "auth", "activate-service-account", "--key-file=" + gcloudAuthFilePath)
+		if _, err := cmd.Output(); err != nil {
+			return fmt.Errorf("Failed to activate gcloud service account: %v", err)
+		}
+	}
+	return nil
+}
+
+// Create logfile for systemd service in outputDir with the given journalctl outputMode.
+func createSystemdLogfile(service string, outputMode string, outputDir string) error {
+	// Generate the journalctl command.
+	var journalCmd string
+	if service == "kern" {
+		journalCmd = fmt.Sprintf("sudo journalctl --output=%v -k", outputMode)
+	} else {
+		journalCmd = fmt.Sprintf("sudo journalctl --output=%v -u %v.service", outputMode, service)
+	}
+	cmd := exec.Command("/bin/sh", "-c", journalCmd)
```

Comment:
This one is a tricky issue and I failed to find any solution so far. Even the current command (with or without sudo) fails to run. The problem is with using ubuntu base image. As pointed out by @Random-Liu (who also previously tried using journalctl for npd):
The journald in ubuntu compiled with incompatible compile configuration with GCI.
```
Journal file /var/log/journal/0e429019876a34f9c55ec02a99498747/system.journal uses an unsupported feature, ignoring file.
-- No entries --
```

In the case of NPD however, we could get away by using fedora, for which journalctl works fine. However for our use, we also need to install gcloud and it fails on fedora due to python issue.
I'm trying to make gcloud work on fedora currently (with no success so far). Is anyone aware of a base image that could be useful here?